tests/structured_conf/data/attr_classes.py

```diff
@@ -492,3 +487,39 @@ class UntypedList:
 class UntypedDict:
     dict: Dict = {"foo": "var"}  # type: ignore
     opt_dict: Optional[Dict] = None  # type: ignore
+
+
+@attr.s(auto_attribs=True)
+class Book:
+    author: Union[str, List[str]] = "dude"
+
+
+@attr.s(auto_attribs=True)
+class Shelf:
+    content: Union[Book, int] = 1
+
+
+@attr.s(auto_attribs=True)
+class Shelf2:
+    content: Union[Book, List[Book]] = Book()
+
+
+@attr.s(auto_attribs=True)
+class OptionalBook:
+    author: Optional[Union[str, List[str]]] = "author"
+
+
+@attr.s(auto_attribs=True)
+class ListUnion:
+    list: List[Union[float, int]] = [1, 3.14]
+
+
+@attr.s(auto_attribs=True)
+class DictUnion:
+    dict: Dict[str, Union[float, int]] = {"foo": 1}
+
+
+@attr.s(auto_attribs=True)
+class UnionWithContainer:
+    union_dict: Union[DictConfig, int] = DictConfig({"foo": 1})
+    union_list: Union[ListConfig, int] = ListConfig([1, 2])
```

Comment:
Add a test for a case for dataclasses as union elements sharing a common base class.
It's weird but people are doing it.

```python
@dataclass
class Base: ...

@dataclass
class Subclass1(Base): ...

@dataclass
class Subclass2(Base): ...

@dataclass
class UnionOfSubclasses:
    foo: Union[Subclass1, Subclass2] 
```